File: crates/adapters/inbound/powehi-rest-api/src/routes/media.rs
//! Media pre-signed URL endpoints.
//!
//! Security invariant: the server NEVER receives ciphertext bytes. Clients
//! upload/download directly to/from R2 using short-lived pre-signed URLs.
//!
//! POST  /v1/media/upload-url        — allocate MediaId + get a pre-signed PUT URL
//! POST  /v1/media/:id/confirm       — mark upload complete (metadata only)
//! GET   /v1/media/:id/download-url  — get a pre-signed GET URL (uploader OR group member)
//! DELETE /v1/media/:id              — delete (uploader device only)

use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use powehi_domain::{group::GroupId, media::MediaId};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{error::ApiError, middleware::AuthenticatedDevice, AppState};

/// 100 MB hard cap on media uploads. Must match `powehi_r2::MAX_MEDIA_BYTES`.
const MAX_UPLOAD_BYTES: u64 = 100 * 1024 * 1024;

#[derive(Deserialize)]
pub struct UploadRequest {
    content_type: String,
    size_bytes: u64,
    /// MLS group this blob is being shared to. When set, any group member may download.
    group_id: Option<Uuid>,
}

#[derive(Serialize)]
pub struct UploadResponse {
    media_id: Uuid,
    upload_url: String,
}

pub async fn request_upload(
    State(state): State<AppState>,
    AuthenticatedDevice(device_id): AuthenticatedDevice,
    Json(body): Json<UploadRequest>,
) -> Result<Json<UploadResponse>, ApiError> {
    if body.size_bytes == 0 || body.size_bytes > MAX_UPLOAD_BYTES {
        return Err(ApiError::from(
            powehi_domain::error::DomainError::InvalidInput("size_bytes out of range".into()),
        ));
    }
    let group_id = body.group_id.map(GroupId::from);
    let (media_id, url) = state
        .media
        .request_upload(
            &device_id,
            &body.content_type,
            body.size_bytes,
            group_id.as_ref(),
        )
        .await
        .map_err(ApiError::from)?;
    Ok(Json(UploadResponse {
        media_id: media_id.as_uuid(),
        upload_url: url,
    }))
}

pub async fn confirm_upload(
    State(state): State<AppState>,
    AuthenticatedDevice(device_id): AuthenticatedDevice,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, ApiError> {
    let media_id = MediaId::from(id);
    state
        .media
        .confirm_upload(&media_id, &device_id)
        .await
        .map_err(ApiError::from)?;
    Ok(StatusCode::NO_CONTENT)
}

#[derive(Serialize)]
pub struct DownloadUrlResponse {
    download_url: String,
}

pub async fn get_download_url(
    State(state): State<AppState>,
    AuthenticatedDevice(device_id): AuthenticatedDevice,
    Path(id): Path<Uuid>,
) -> Result<Json<DownloadUrlResponse>, ApiError> {
    let media_id = MediaId::from(id);
    let url = state
        .media
        .get_download_url(&media_id, &device_id)
        .await
        .map_err(ApiError::from)?;
    Ok(Json(DownloadUrlResponse { download_url: url }))
}

pub async fn delete_media(
    State(state): State<AppState>,
    AuthenticatedDevice(device_id): AuthenticatedDevice,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, ApiError> {
    let media_id = MediaId::from(id);
    state
        .media
        .delete(&media_id, &device_id)
        .await
        .map_err(ApiError::from)?;
    Ok(StatusCode::NO_CONTENT)
}

File: app/src/api/media.ts
/**
 * Media presigned-URL API client (prd.md §9.2).
 *
 * Security invariants:
 * - The server NEVER receives plaintext or ciphertext bytes — clients upload/
 *   download directly to/from Cloudflare R2 using short-lived presigned URLs.
 * - Bearer token is always in the Authorization header, never in the URL.
 * - All requests use HTTPS (enforced by the browser for Bearer-auth endpoints).
 */

const API_BASE = "/v1";

function authHeaders(token: string): HeadersInit {
	return {
		"Content-Type": "application/json",
		Authorization: `Bearer ${token}`,
	};
}

async function throwOnError(resp: Response): Promise<void> {
	if (!resp.ok) {
		const body = (await resp.json().catch(() => ({}))) as { code?: string };
		throw new Error(body.code ?? `http_${resp.status}`);
	}
}

/**
 * POST /v1/media/upload-url — allocate a MediaId and get a presigned R2 PUT URL.
 *
 * @param token        Bearer session token
 * @param contentType  MIME type (e.g. "image/jpeg"). The server stores metadata only.
 * @param sizeBytes    Exact byte count of the encrypted blob (1–100MB).
 * @param groupId      Optional MLS group UUID. When set, all group members may download.
 * @returns            { mediaId, uploadUrl } — PUT ciphertext to uploadUrl directly.
 */
export async function requestMediaUpload(
	token: string,
	contentType: string,
	sizeBytes: number,
	groupId?: string,
): Promise<{ mediaId: string; uploadUrl: string }> {
	const body: Record<string, unknown> = {
		content_type: contentType,
		size_bytes: sizeBytes,
	};
	if (groupId !== undefined) {
		body.group_id = groupId;
	}
	const resp = await fetch(`${API_BASE}/media/upload-url`, {
		method: "POST",
		headers: authHeaders(token),
		body: JSON.stringify(body),
	});
	await throwOnError(resp);
	return resp.json() as Promise<{ mediaId: string; uploadUrl: string }>;
}

/**
 * POST /v1/media/:id/confirm — mark an upload complete after the R2 PUT succeeds.
 *
 * @param token    Bearer session token
 * @param mediaId  UUID returned by requestMediaUpload
 */
export async function confirmMediaUpload(token: string, mediaId: string): Promise<void> {
	const resp = await fetch(`${API_BASE}/media/${mediaId}/confirm`, {
		method: "POST",
		headers: authHeaders(token),
	});
	await throwOnError(resp);
}

/**
 * GET /v1/media/:id/download-url — get a presigned R2 GET URL.
 *
 * The caller must be either the uploader or a member of the group the blob was
 * shared to. Fetch the blob from downloadUrl directly; the server never sees
 * the ciphertext bytes.
 *
 * @param token    Bearer session token
 * @param mediaId  UUID of the blob
 * @returns        { downloadUrl } — GET ciphertext from this URL, then decrypt.
 */
export async function getMediaDownloadUrl(
	token: string,
	mediaId: string,
): Promise<{ downloadUrl: string }> {
	const resp = await fetch(`${API_BASE}/media/${mediaId}/download-url`, {
		headers: authHeaders(token),
	});
	await throwOnError(resp);
	return resp.json() as Promise<{ downloadUrl: string }>;
}

/**
 * DELETE /v1/media/:id — delete a blob (uploader device only).
 *
 * @param token    Bearer session token
 * @param mediaId  UUID of the blob to delete
 */
export async function deleteMedia(token: string, mediaId: string): Promise<void> {
	const resp = await fetch(`${API_BASE}/media/${mediaId}`, {
		method: "DELETE",
		headers: authHeaders(token),
	});
	await throwOnError(resp);
}
